### ISSUE TITLE
Allow linkerd-init initContainer injection to be configurable via workload annotation

### ISSUE THREAD

**Issue by aw185176** [linkerd/linkerd2#7999]
### What problem are you trying to solve?

Meshing databases while also meshing the initContainers that work against them (https://github.com/linkerd/linkerd2/issues/3812).

### How should the problem be solved?

Allow injection of the linkerd-init initContainer to optionally be done at the front of the array.

### Any alternatives you've considered?

We are going to shuffle our initContainers into the app itself, but considering this is a bog standard use case for initContainers, and the KEPs linked from #3812 were rejected, it seems reasonable that Linkerd support this use case.

Otherwise, we don't have much of a choice other than to weaken our mesh's security stance.

### How would users interact with this feature?

A `config.linkerd[...]` annotation seems reasonable.

### Would you like to work on this feature?

no

**Comment by olix0r** [comment]
Does this need to be configurable? Would it be better to have the init container always be added first?

**Comment by krzysztofdrys** [comment]
I would like to work on this. 

Right now linkerd init container is added last. I hope that adding it first should solve the issue. I think it should be enough to do this change (as per [JSON Patch spec](https://datatracker.ietf.org/doc/html/rfc6902#page-12)): 

```
diff --git a/charts/patch/templates/patch.json b/charts/patch/templates/patch.json
index fd3697ee..4c3fabf1 100644
--- a/charts/patch/templates/patch.json
+++ b/charts/patch/templates/patch.json
@@ -62,7 +62,7 @@
   },
   {
     "op": "add",
-    "path": "{{$prefix}}/spec/initContainers/-",
+    "path": "{{$prefix}}/spec/initContainers/0",
     "value":
       {{- include "partials.proxy-init" . | fromYaml | toPrettyJson | nindent 6 }}
```

However, I would like to add some tests before submitting the PR.

**Comment by krzysztofdrys** [comment]
I have started on working on change which puts linkerd init container always first: https://github.com/linkerd/linkerd2/pull/8034/files

However, I don't really understand why it should help in this case (db migration). From what I understand:
- init containers are run one-by-one, from first to last (https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)
- migration container is no able to make connection, because proxy sidecar is not running.

So TBH I don't see how putting linkerd init container first can help here. Nevertheless, maybe this PR can be useful anyway?

**Comment by krzysztofdrys** [comment]
Also: if this should be behind config annotation, I will be happy to add it.

**Comment by mateiidavid** [comment]
@krzysztofdrys I think it's a sensible default, if OP needs this to be configurable we can possibly discuss it further. In regards to making connections in an init container, there's always `--skip-<inbound/outbound>--ports` to bypass the proxy.

**Comment by krzysztofdrys** [comment]
Good! Then I have a PR ready.

**Comment by aw185176** [comment]
>migration container is no able to make connection, because proxy sidecar is not running.

Good call out.  Part of the problem is that the certificate doesn't exist (because `linkerd-init` has not requested it yet).  The other issue that I totally missed when working through this is that the proxy sidecar still won't be running.  I do think that this will enable workarounds such as the initContainer leveraging the cert directly to create its own TLS connection to the database, but I could be missing something.

**Comment by aw185176** [comment]
>Does this need to be configurable? Would it be better to have the init container always be added first?

I agree that initContainer first would be better for more scenarios than the current behavior.  Based on the issues I linked, it seemed like there might be some use case that I am not aware of (or existing expectations) that I didn't want to rule out with my request.

>@krzysztofdrys I think it's a sensible default, if OP needs this to be configurable we can possibly discuss it further. In regards to making connections in an init container, there's always --skip-<inbound/outbound>--ports to bypass the proxy.

Agree, I don't think we actually need it to be configurable.  Feel free to close this issue when the PR lands.  

Thanks all for the feedback and sensible discussion

**Comment by kleimkuhler** [comment]
After thinking more about this change it may actually be more harmful to inject Linkerd's init container as the first init container rather than the last. If it is the first init container, then all init containers that run after it will go through the iptables rules that were re-written by it.

As it is — being injected as the last init container — there are no issues in most situations. The already existing init containers run to completion with the unmodified iptables; because Linkerd's init container is the last to run we only start encountering these modified iptables rules when the proxy and application starts up.

Unfortunately, this does mean there isn't really a solution to this right now aside from using the `skip-{inbound, outbound}-ports` annotation. This should solve issues with init containers completing, but it does sometimes mean that the application's connections to the same ports are not included in the mesh.

**Comment by krzysztofdrys** [comment]
I understand and share these concerns. I would like to stress that however this will be resolved, I am happy that I implemented https://github.com/linkerd/linkerd2/pull/8034 Even it gets closed. :-) 

In my opinion, the best default option would be to make Linkerd init container first. Linkerd makes a promise to make all connections meshed. In the case of init containers, this promise cannot be fulfilled, because of k8s limitations. But at least we make Linkerd users aware of this, by forcing them to write `skip-{inbound, outbound}-ports`.

BUT, if we make this change right now, it might break some people's setup. I would hate this as a user. IMO https://github.com/linkerd/linkerd2/pull/8034  should not be merged as it is now.

What are our options?

1. Leave it as it is and just close https://github.com/linkerd/linkerd2/pull/8034 I am fine with that because I already had my share of fun implementing it.

2. Add an optional annotation to control if linkerd container goes first or last and default it to last. This gives us backward compatibility. The drawback here is that we are adding a switch that no one will ever use.

3. Add an optional annotation to control if linkerd container goes first or last and default it to first. But also make "first", when migrating from an existing cluster. This way we have backward compatibility and also we "force" new users to use a better setting. The drawback here is that we are adding complexity to solve a problem that might not be that important after all. 

I see these three options. If (2) or (3) gets selected, I would be happy to implement it.

**Comment by mikebell90** [comment]
Pardon my ignorance but

- You can never guarantee initContainer ordering. That in itself assumes it is the last mutable webhook controller. While IIRC there's a deterministic alphabetical thing, that's fragile, non obvious, and not really meant to be depended on. So you cannot guarantee the order of initContainer injection (the same honestly applies to linkerd's attempt to force the matter with containers. I'm not fond of "90% but oops" solutions myself ;)
- I could be missing something, but the issue isn't the initContainers (unless one uses the DB in an initContainer, in which case yes the iptables issue occurs - this is also true of Vault etc or anything else running in initContainers and is a fundamental draw back of the whole approach, but the appContainers which isn't really addressed until the century a variant of the sidecar KEPS is approved.

**Comment by krzysztofdrys** [comment]
> You can never guarantee initContainer ordering.

I don't think this is quite true. As per [spec](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/):

> During Pod startup, the kubelet delays running init containers until the networking and storage are ready. Then the kubelet runs the Pod's init containers in the order they appear in the Pod's spec.

**Comment by mikebell90** [comment]
Sure I mean the final form of the yaml given multiple mutating webhook controllers. This is not deterministic in the sense that it will change very easily and fragile - can you promise you will never add or substract any admission controllers? Never install another app that might contain one ? Probably not hence you’ll have a lot of trouble avoiding regression

**Comment by jault3** [comment]
> If it is the first init container, then all init containers that run after it will go through the iptables rules that were re-written by it.

I would agree with what @kleimkuhler mentions here and a few others mentioned above. Running the Linkerd init container first essentially removes all network access from subsequent init containers because those iptables rules would route the traffic to a sidecar that isn't running. If that was worked around somehow, it would also make using third party init containers that need network access more difficult, such as the [vault agent injector](https://www.vaultproject.io/docs/platform/k8s/injector). I would foresee no longer being able to use that without modifications to use the TLS cert manually.

FWIW, I also ran in to this issue. One of the reasons I can't start using authorization policies is because of database migrations. There are efforts to have the applications move migrations into the main source code and run as a first task on startup rather than an init container. But without authorization policies, the non-meshed init container can talk to the meshed database without issues.

**Comment by kleimkuhler** [comment]
After the above discussion, the issue that this originally pertained to—which was allowing `linkerd-init` container ordering to be configurable—can be closed. Due to the rewriting of IP tables by `linkerd-init`, it does not seem desirable to ever have this container run _before_ other init containers since they would then have their traffic rerouted through the proxy which will not yet be running.

Injecting `linkerd-init` as the last init container (last currently, another webhook may add more) is the best that we can do right now.

With regard to policy, this is being worked on in other issues and should hopefully allow for a smoother experience of allowing other init containers to communicate properly.
